### PR TITLE
Refactor usages of `physicalSizeTestValue` to simplify code and improve DX

### PR DESCRIPTION
**Pull request by fzyzcjy**
Just a small refactor. Using:

```dart
  set physicalSizeCurrentTestValue(ui.Size value) {
    physicalSizeTestValue = value;
    addTearDown(clearPhysicalSizeTestValue);
  }
```

We can make setting physical size test values a bit better in the following two aspects:

1. Code is less duplicated. Originally need to specify set value + clear value, now only need one call
2. Writing new tests are less error-prone, especially for new learners of Flutter, thus increasing DX. When I firstly learn Flutter, I often wrongly write down set test value calls, *without* the clear value calls. Then the tests works pretty well when isolated, but you know, it fails weirdly when run sequentially because the later tests have wrong physical size. As a new learner of Flutter (years ago), it took me some time before realizing it is this bug. Thus, it would be great to avoid the possibility of such problem from the beginning.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter_test/lib/src/window.dart [hunk 1]
```diff
@@ -518,6 +519,20 @@ class TestWindow implements ui.SingletonFlutterWindow {
   }
 }
 
+/// Convenient methods for auto reset test values
```
Comment (justinmc) [inline packages/flutter_test/lib/src/window.dart:522]: "Convenience method for automatically resetting test values."

File: packages/flutter_test/lib/src/window.dart [hunk 2]
```diff
@@ -518,6 +519,20 @@ class TestWindow implements ui.SingletonFlutterWindow {
...
+// Use an `extension`, such that classes implementing TestWindow will not be affected
```
Comment (justinmc) [inline packages/flutter_test/lib/src/window.dart:523]: `//` should be `///`.  Also there should be an empty line of `///` above this line.

Comment (justinmc) [reply packages/flutter_test/lib/src/window.dart:523]: "Uses an `extension` so that classes implementing TestWindow will not be affected."

File: packages/flutter_test/lib/src/window.dart [hunk 3]
```diff
@@ -518,6 +519,20 @@ class TestWindow implements ui.SingletonFlutterWindow {
...
+extension TestWindowExtensionValueTearDown on TestWindow {
```
Comment (justinmc) [inline packages/flutter_test/lib/src/window.dart:524]: The Flutter style guide [says](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#tldr) to avoid `extension`.  Is there any way to improve this without using `extension`?

**Review by justinmc — CHANGES_REQUESTED** [review]
I think you're right that it would be nice to not have to worry about cleaning up after setting `physicalSizeTestValue` etc., but I don't know if we can make this kind of change.

Firstly see my comment about `extension`, which we try to avoid.

Secondly, I think having two similarly named setters, one that cleans up after itself and one that doesn't, is still pretty unintuitive for developers.  There is still a lot of potential for making a mistake, say by accidentally setting `physicalSizeTestValue` instead of  `physicalSizeCurrentTestValue` and expecting it to automatically do `clearPhysicalSizeTestValue`.

Maybe there is another approach that would improve this situation, but I'm not sure what it is.

File: packages/flutter_test/lib/src/window.dart (hunk 3 above)
Comment (fzyzcjy) [reply packages/flutter_test/lib/src/window.dart:524]: By implementing it directly into the TestWindow. This has the slight drawback that anyone who `implements` TestWindow will need to add one more method, but if few people are doing so then looks ok

**Comment by fzyzcjy** [comment]
> Secondly, I think having two similarly named setters, one that cleans up after itself and one that doesn't, is still pretty unintuitive for developers. There is still a lot of potential for making a mistake, say by accidentally setting physicalSizeTestValue instead of physicalSizeCurrentTestValue and expecting it to automatically do clearPhysicalSizeTestValue.

What about renaming it to: physicalSizeTestValue vs physicalSizeTestValueAutoClear? Then nobody can use it by mistake since the latter says "auto clear"

File: packages/flutter_test/lib/src/window.dart [hunk 4]
```diff
@@ -81,6 +82,12 @@ class TestWindow implements ui.SingletonFlutterWindow {
     onMetricsChanged?.call();
   }
 
+  /// Just [devicePixelRatioTestValue], but automatically reset after the test finishes
```
Comment (Renzo-Olivares) [inline packages/flutter_test/lib/src/window.dart:85]: Nit: maybe `/// Sets the [devicePixelRatioTestValue], and automatically resets after the test finishes...`

File: packages/flutter_test/lib/src/window.dart [hunk 5]
```diff
@@ -97,6 +104,12 @@ class TestWindow implements ui.SingletonFlutterWindow {
     onMetricsChanged?.call();
   }
 
+  /// Just [physicalSizeTestValue], but automatically reset after the test finishes
```
Comment (Renzo-Olivares) [inline packages/flutter_test/lib/src/window.dart:107]: Nit: see similar comment above.

File: packages/flutter_test/lib/src/window.dart [hunk 6]
```diff
@@ -81,6 +82,12 @@ class TestWindow implements ui.SingletonFlutterWindow {
     onMetricsChanged?.call();
   }
 
+  /// Just [devicePixelRatioTestValue], but automatically reset after the test finishes
+  set devicePixelRatioTestValueAutoClear(double value) { // ignore: avoid_setters_without_getters
```
Comment (Renzo-Olivares) [inline packages/flutter_test/lib/src/window.dart:86]: Add a doc reference to this new auto clear method in the docs for `devicePixelRatioTestValue`.

File: packages/flutter_test/lib/src/window.dart [hunk 7]
```diff
@@ -97,6 +104,12 @@ class TestWindow implements ui.SingletonFlutterWindow {
     onMetricsChanged?.call();
   }
 
+  /// Just [physicalSizeTestValue], but automatically reset after the test finishes
+  set physicalSizeTestValueAutoClear(ui.Size value) { // ignore: avoid_setters_without_getters
```
Comment (Renzo-Olivares) [inline packages/flutter_test/lib/src/window.dart:108]: Add a doc reference to this new auto clear method in the docs for `physicalSizeTestValue`.

**Review by Renzo-Olivares — CHANGES_REQUESTED** [review]
This seems like a reasonable change to make. Please see my comments, mostly related to documentation. Adding a small tidbit about these new auto clear setters to the public non-auto clear setters would be great for developers to discover this method.

File: packages/flutter_test/lib/src/window.dart (hunk 4 above)
Comment (justinmc) [inline packages/flutter_test/lib/src/window.dart:85]: Period at the end.

File: packages/flutter_test/lib/src/window.dart (hunk 5 above)
Comment (justinmc) [inline packages/flutter_test/lib/src/window.dart:107]: Period at the end.

**Review by justinmc — COMMENTED** [review]
Thanks, the "AutoClear" name and the lack of `extension` make this a lot better to me.

**Comment by pdblasi-google** [comment]
I like the goal of this (easier tests, less likely to make mistakes), but I have a couple of concerns with this as implemented:

1. Does this _hide_ a problem, rather than addressing it?
    * It's pretty common that `addTearDown` and `tearDown` calls are missed as you pointed out in your description. This PR makes it easier to not miss those calls for a few specific cases, but it doesn't really address _why_ those calls are missed or drive users to write better tests in other cases.
2. Does adding these methods here set a standard that the testing framework will be expected to stick to elsewhere?
    * How do we determine which properties or fields should/shouldn't have versions that auto clear?
    * How should adding those auto clear versions be prioritized?
    * If/when those versions break, should it be considered a breaking change and should the breaking change process be followed for just the auto clear version, or for auto clear and standard versions?
3. Will this cause unintuitive behavior in other use cases?
    * Specifically I'm thinking of someone trying to use this for setting up a `group` of tests. Since this makes use of `addTearDown`, it'll only behave as expected for individual tests. If someone tries to use the "AutoClear" version in a group, it'll throw an exception.
    * In combination with number 2, would any "AutoClear" versions also need an "AutoClearGroup" version?

Personally, I'd lean towards pushing the convenience method part to a separate package where the expectations for support, maintenance, and coverage won't be as high. It'd also allow for more community contributions for different conveniences and extensions to be added without needing to go through as much process as this repo requires. As for the fixes to our own tests, it's a great catch and should be included, but IMO without using the convenience method.

**Comment by fzyzcjy** [comment]
@Renzo-Olivares @justinmc done :)

**Comment by fzyzcjy** [comment]
@pdblasi-google 

> Does this hide a problem, rather than addressing it?
It's pretty common that addTearDown and tearDown calls are missed as you pointed out in your description. This PR makes it easier to not miss those calls for a few specific cases, but it doesn't really address why those calls are missed or drive users to write better tests in other cases.

So what do you think is the reason "why those calls are missed"? I indeed am not sure - maybe because they just forget?

> Does adding these methods here set a standard that the testing framework will be expected to stick to elsewhere?
How do we determine which properties or fields should/shouldn't have versions that auto clear?

For fields that have a `set somethingTestValue` and `clearSomethingTestValue` I guess?

> How should adding those auto clear versions be prioritized?

IMHO this is quite trivial to implement (if you like I can submit more)

> If/when those versions break, should it be considered a breaking change and should the breaking change process be followed for just the auto clear version, or for auto clear and standard versions?

Sorry, not sure about the question

> Will this cause unintuitive behavior in other use cases?
Specifically I'm thinking of someone trying to use this for setting up a group of tests. Since this makes use of addTearDown, it'll only behave as expected for individual tests. If someone tries to use the "AutoClear" version in a group, it'll throw an exception.
In combination with number 2, would any "AutoClear" versions also need an "AutoClearGroup" version?

Luckily, not a problem! If you like I can put the following code which automatically uses addTearDown vs tearDown and is used internally in my app:

```dart
void autoSetUpTearDownOrAddTearDownSync(
  void Function() setUpBody,
  void Function() tearDownBody,
) {
  // ref: test_api :: test_structure.dart :: addTearDown
  // `Invoker.current==null` will lead to a message `addTearDown() may only be called within a test.`
  // so use this to determine whether in test
  if (Invoker.current == null) {
    setUp(setUpBody);
    tearDown(tearDownBody);
  } else {
    setUpBody();
    addTearDown(tearDownBody);
  }
}
```

**Comment by pdblasi-google** [comment]
@fzyzcjy

Personally, I think those calls are missed because they aren't made obvious to users through documentation and examples.  As such, I'm not sure hiding those calls behind convenience methods promotes better usage of those methods.

Adding the convenience methods _also_ increases our API surface, which increases our maintenance burden, and sets up expectations for other parts of the testing framework that don't actually _solve_ the problem of those methods being missed, but instead makes it even less likely that users will run into an example using them correctly.

Long story short, I don't think the convenience methods are worth the long term cost of introducing that pattern to the framework. Hence suggesting creating a package for them, where the maintenance costs and coverage expectations are going to be lower.

It's also worth noting that there's work at the moment to move from a single window to multiple views (to support desktop better), which may be a great opportunity to improve on these particular fields' APIs from the ground up. I believe @goderbauer is driving that initiative, though he's out until next Tuesday.

**Comment by fzyzcjy** [comment]
@pdblasi-google 

> Personally, I think those calls are missed because they aren't made obvious to users through documentation and examples.

Not sure how other people do, but I am not this case personally. I have this simply because I forgot to call the clear :/ I do vaguely remember the clear, and if you test me in a questionaire I will remember it, but when programming I just set the value - doing the minimal necessary work - while forgetting the clear.

> Hence suggesting creating a package for them

Then what should we do for the flutter framework itself? Or shall we mark them `@internal` so flutter framework can use it?

> It's also worth noting that there's work at the moment to move from a single window to multiple views (to support desktop better), which may be a great opportunity to improve on these particular fields' APIs from the ground up. I believe @goderbauer is driving that initiative, though he's out until next Tuesday.

Looks interesting!

**Comment by pdblasi-google** [comment]
If the convenience methods are moved to a package, I would suggest just not using them in the flutter repo and updating the tests here to use the `addTearDown` methods where appropriate to clear the values. That way, those tests can serve as an example of how to use the standard `addTearDown` pattern.

**Comment by fzyzcjy** [comment]
> not using them in the flutter repo*

Then the tests are not DRY ;)

**Comment by pdblasi-google** [comment]
> Then the tests are not DRY ;)

I mean, if you want to get the tests _really_ DRY, then the setup/teardown would be moved out to the `group` level and done either once for the entire group, or set up to happen before and after each test with the group level methods.

Since we're working on a public API though, we don't necessarily _want_ to be DRY on everything. We have to balance trying to keep our internal stuff as clean as we can with not making the API too specific to our internal use cases.

**Comment by fzyzcjy** [comment]
> then the setup/teardown would be moved out to the group level and done either once for the entire group

Well, those tests are not in the same file, so IMHO they cannot be solved by `group`.

> Since we're working on a public API though, we don't necessarily want to be DRY on everything. We have to balance trying to keep our internal stuff as clean as we can with not making the API too specific to our internal use cases.

I see. Btw, it is not "too specific to our internal use cases" as it is used everyday by external users

**Comment by justinmc** [comment]
I have to agree with @pdblasi-google, I'm not certain that this is better than the existing API, so I don't think we should make this change.  Instead, here's what I think we can do about this problem:

  1. Clearing should happen immediately after setting.  This makes it obvious to future readers of the test that these values need to be cleared, and it makes it less likely the test will end without the values being cleared.  I think this could be done in this PR or in a new PR.

Old way:

```dart
tester.binding.window.devicePixelRatioTestValue = 1.2;
tester.binding.window.physicalSizeTestValue = const Size(250, 300);

... // Rest of test

 tester.binding.window.clearPhysicalSizeTestValue();
 tester.binding.window.clearDevicePixelRatioTestValue();
```

Should be changed to:

```dart
tester.binding.window.devicePixelRatioTestValue = 1.2;
addTearDown(clearDevicePixelRatioTestValue);
tester.binding.window.physicalSizeTestValue = const Size(250, 300);
addTearDown(clearPhysicalSizeTestValue);

... // Rest of test
```

  2. `devicePixelRatioTestValueAutoClear` and `physicalSizeTestValueAutoClear` could be released as a Pub package if desired.

  3.  Like @pdblasi-google said, ideas about autoclear and other patterns that avoid this problem should be directed at @goderbauer's work to move from TestWindow to TestView, since we have the opportunity to start fresh here and change the API.  He has a [design doc](https://docs.google.com/document/d/1Z7Qrb08dOnfB8IxPgsyujVn4MsDcYPUUku9CeF70_S0/edit), which it doesn't specifically mention this about testing so it might be worth bringing it up in a comment there.

**Comment by fzyzcjy** [comment]
Hmm I see. Then maybe I will refactor like what you said.

P.S Did a comment there just now.

**Comment by fzyzcjy** [comment]
Getting back to this PR. It looks like we are already using `addTearDown(tester.view.reset)` in the master. So what should I do now :)

image

**Comment by pdblasi-google** [comment]
As a note, I have an issue (#121917) that'll handle the more generic case. Not sure when I'll get to it, but it should cover the fixes that were worked on here originally as well when I do.

**Comment by fzyzcjy** [comment]
I think so. Just wanted to hear your opinions :)

**Comment by justinmc** [comment]
Sounds good, thanks for helping us figure this out even though there's nothing to change here!

**Comment by fzyzcjy** [comment]
You are welcome!
